### PR TITLE
Remove event listeners after exporting/importing a site

### PR DESCRIPTION
**Pull request by fluiddot**
Related to https://github.com/Automattic/studio/pull/370#discussion_r1689539008.

## Proposed Changes

- Return a function in `handleEvents` to remove all listeners created.
- Remove listeners in the various calls to `handleEvents` in import/export processes.

## Testing Instructions

- Apply the following patch:
```patch
diff --git forkSrcPrefix/src/lib/import-export/handle-events.ts forkDstPrefix/src/lib/import-export/handle-events.ts
index 33633d3d5933ca19517922c812fef68735c75dbe..913a4ad94a32f6419e34cabb206e4bb0d8f115d9 100644
--- forkSrcPrefix/src/lib/import-export/handle-events.ts
+++ forkDstPrefix/src/lib/import-export/handle-events.ts
@@ -23,7 +23,10 @@ export const handleEvents = (
 			} );
 		};
 		emitter.on?.( eventName, listener );
-		removeListeners.push( () => emitter.off?.( eventName, listener ) );
+		removeListeners.push( () => {
+			emitter.off?.( eventName, listener );
+			console.log( 'Listener removed for event', eventName );
+		} );
 	} );
 	return () => {
 		removeListeners.forEach( ( remove ) => remove() );

```
- Follow instructions from https://github.com/Automattic/studio/pull/370.
- After importing a site, observe logs with the message `Removing listeners`.
- After exporting a site, observe logs with the message `Removing listeners`.

## Pre-merge Checklist

- [x] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/lib/import-export/import/import-manager.ts [hunk 1]
```diff
@@ -37,21 +38,28 @@ export async function importBackup(
...
+			const result = await importer.import( sitePath );
+			removeImportListeners();
+			return result;
 		} else {
 			throw new Error( 'No suitable importer found for the given backup file' );
 		}
 	} catch ( error ) {
 		console.error( 'Backup import failed:', ( error as Error ).message );
+		removeBackupListeners?.();
+		removeImportListeners?.();
```
Comment (sejas) [inline src/lib/import-export/import/import-manager.ts]: What do you think about moving them to the finally closure?
We could potentially move all the remove listeners there.

**Review by sejas — APPROVED** [review]
I think the patch on the description should be something like:

```diff
diff --git a/src/lib/import-export/types.ts b/src/lib/import-export/types.ts
index 33633d3..1e54a87 100644
--- a/src/lib/import-export/types.ts
+++ b/src/lib/import-export/types.ts
@@ -23,7 +23,10 @@ export const handleEvents = (
                        } );
                };
                emitter.on?.( eventName, listener );
-               removeListeners.push( () => emitter.off?.( eventName, listener ) );
+               removeListeners.push( () => {
+                       emitter.off?.( eventName, listener );
+                       console.log( 'removed listener', eventName );
+               } );
        } );
        return () => {
                removeListeners.forEach( ( remove ) => remove() );
```

After applying my patch, I confirm I see the log removing the listeners in the console of NodeJS.

Screenshot 2024-07-24 at 16 07 37

File: src/lib/import-export/import/import-manager.ts (hunk 1 above)
Comment (fluiddot) [reply src/lib/import-export/import/import-manager.ts]: Yeah, good idea. We can rely on the `finally` for removing all listeners.

Comment (fluiddot) [reply src/lib/import-export/import/import-manager.ts]: I've added this suggestion in https://github.com/Automattic/studio/pull/383/commits/a9fefac4fc6af632883f72158167f23acb7fddca.

**Comment by fluiddot** [comment]
> I think the patch on the description should be something like:
> 
> ```diff
> diff --git a/src/lib/import-export/types.ts b/src/lib/import-export/types.ts
> index 33633d3..1e54a87 100644
> --- a/src/lib/import-export/types.ts
> +++ b/src/lib/import-export/types.ts
> @@ -23,7 +23,10 @@ export const handleEvents = (
>                         } );
>                 };
>                 emitter.on?.( eventName, listener );
> -               removeListeners.push( () => emitter.off?.( eventName, listener ) );
> +               removeListeners.push( () => {
> +                       emitter.off?.( eventName, listener );
> +                       console.log( 'removed listener', eventName );
> +               } );
>         } );
>         return () => {
>                 removeListeners.forEach( ( remove ) => remove() );
> ```
> 
> After applying my patch, I confirm I see the log removing the listeners in the console of NodeJS.

Good catch @sejas. I've updated the patch with the proper instructions.

**Pull request merged**